src/Workspaces/Core/Portable/Utilities/BKTree.cs

```diff
@@ -0,0 +1,184 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Roslyn.Utilities;
+using static System.Math;
+using static Roslyn.Utilities.PortableShim;
+
+namespace Roslyn.Utilities
+{
+    /// <summary>
+    /// NOTE: do not use this class directly.  It is intended for use only by the spell checker.
+    /// </summary>
+    internal partial class BKTree
```

Comment:
For the EditDistance, you cited the paper and explained the general idea. A similar thing would be useful for this less-widely-known data structure.